### PR TITLE
Okta and AWS broader error handling

### PR DESCRIPTION
**Pull request by alex-fang0**
### Why
Adding more error handling to network calls to avoid the general "BMX exited with unexpected internal error" error.

### CODE REVIEW

File: src/D2L.Bmx/Aws/AwsClient.cs [hunk 1]
```diff
@@ -13,18 +13,26 @@ internal class AwsClient( IAmazonSecurityTokenService stsClient ) : IAwsClient {
...
+
+			return new AwsCredentials(
+				SessionToken: authResp.Credentials.SessionToken,
+				AccessKeyId: authResp.Credentials.AccessKeyId,
+				SecretAccessKey: authResp.Credentials.SecretAccessKey,
+				Expiration: authResp.Credentials.Expiration.ToUniversalTime()
+			);
+		} catch ( Exception ex ) {
+			if (ex.Message == "The requested DurationSeconds exceeds the MaxSessionDuration set for this role.") {
+				throw new BmxException( "Duration exceeds the MaxSessionDuration for this role. Lower it in config/parameter", ex );
```
Comment (gord5500) [inline src/D2L.Bmx/Aws/AwsClient.cs:33]: Do you know what the other errors are and are they just not useful or easy to understand?

Comment (alex-fang0) [reply src/D2L.Bmx/Aws/AwsClient.cs:33]: Yea I looked over the [rest of the errors](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRoleWithSAML.html#API_AssumeRoleWithSAML_Errors) and they seem pretty unlikely to happen and not too user friendly if I we're to add a case

Comment (alex-fang0) [reply src/D2L.Bmx/Aws/AwsClient.cs:33]: Some more general [common errors](https://docs.aws.amazon.com/STS/latest/APIReference/CommonErrors.html) here but most of them seem to be handled by how bmx is used. The only one that could be added I would think is ThrottlingException

Comment (gord5500) [reply src/D2L.Bmx/Aws/AwsClient.cs:33]: Sounds good 🤝

File: src/D2L.Bmx/Okta/OktaApi.cs [hunk 2]
```diff
@@ -92,44 +101,55 @@ internal class OktaApi : IOktaApi {
...
+		try {
+			var resp = await _httpClient.PostAsJsonAsync(
+				"sessions",
+				new CreateSessionRequest( sessionToken ),
+				SourceGenerationContext.Default.CreateSessionRequest );
+
+			var session = await JsonSerializer.DeserializeAsync(
+				await resp.Content.ReadAsStreamAsync(),
+				SourceGenerationContext.Default.OktaSession );
+			return session ?? throw new BmxException( "Creating Octa Session returned null" );
```
Comment (gord5500) [inline src/D2L.Bmx/Okta/OktaApi.cs:136]:
```suggestion
			return session ?? throw new BmxException( "Creating Okta Session returned null" );
```

Also your title has it as Octa

File: src/D2L.Bmx/Okta/OktaApi.cs [hunk 3]
```diff
@@ -49,41 +49,50 @@ internal class OktaApi : IOktaApi {
...
+				StateToken: not null,
+				Embedded.Factors: not null,
+				Status: AuthenticationStatus.MFA_REQUIRED
+			} ) {
+				return new AuthenticateResponse.MfaRequired(
+					authnResponse.StateToken,
+					authnResponse.Embedded.Factors
+				);
+			}
+			throw new BmxException( "Okta API Authentication Response was not Success or MFA Required" );
```
Comment (gord5500) [inline src/D2L.Bmx/Okta/OktaApi.cs:77]: Would this text ever get shown to the user since it's caught a few lines down and then we throw a new exception with different message

Comment (alex-fang0) [reply src/D2L.Bmx/Okta/OktaApi.cs:77]: No this would not show for the user. This is more for when Debug mode is on and it shows in the inner exception to make it a bit easier to figure out what went wrong. Not sure how useful it would be tho

File: src/D2L.Bmx/Aws/AwsClient.cs (hunk 1 above)
Comment (alex-fang0) [reply src/D2L.Bmx/Aws/AwsClient.cs:33]: Looking online I can't even seem to find anyone running into or an example of the ThrottlingException error for AssumeRole. I think it should be fine to leave out

File: src/D2L.Bmx/Okta/OktaApi.cs (hunk 3 above)
Comment (gord5500) [reply src/D2L.Bmx/Okta/OktaApi.cs:77]: Ah Oki. Just wanted to make sure it was planned to not show them

File: src/D2L.Bmx/Okta/OktaApi.cs [hunk 4]
```diff
@@ -92,44 +101,55 @@ internal class OktaApi : IOktaApi {
...
-			?? throw new BmxException( "Error retrieving AWS accounts from Okta" );
+		try {
+			var apps = await _httpClient.GetFromJsonAsync(
+				"users/me/appLinks",
+				SourceGenerationContext.Default.OktaAppArray );
+
+			return apps?.Where( app => app.AppName == "amazon_aws" ).ToArray()
+				?? throw new BmxException( "Retrieving AWS accounts from Okta returned null" );
+		} catch( Exception ex ) {
+			throw new BmxException( "Error retrieving AWS accounts from Okta", ex );
```
Comment (gord5500) [inline src/D2L.Bmx/Okta/OktaApi.cs:151]: I think these extra catches could be misleading. If I change the route [here](https://github.com/Brightspace/bmx/blob/3e365d3e5d48abc8b3f9ec80b3bba868d460d3eb/src/D2L.Bmx/Okta/OktaApi.cs/#L129) from `sessions` to `sessionss` I'd expect to see the exception message mentioning a problem to create an okta session. Instead though I get this message here `Error retrieving AWS accounts from Okta`

Comment (gord5500) [reply src/D2L.Bmx/Okta/OktaApi.cs:151]: Maybe we need to also pass up the caught exception message if it's a bmx exception?

File: src/D2L.Bmx/Okta/OktaApi.cs (hunk 3 above)
Comment (cfbao) [reply src/D2L.Bmx/Okta/OktaApi.cs:77]: I'd really like to avoid nesting BmxExceptions. It's intended as a mechanism to show user facing messages, and I'd like to keep it that way. Using it in multiple different ways makes our code more confusing.

Also, I'd consider throwing exception then immediately catch it below not the best form in general. There's usually a better pattern that can express the desired logic flow.

Comment (boarnoah) [reply src/D2L.Bmx/Okta/OktaApi.cs:77]: Its hard to articulate exactly what "Okta API Authentication Response was not Success or MFA Required" exception is attempting to capture. 

There is a few situations we need to deal with here (in general):
1. Bad initial credentials
1. Failed MFA challenge
1. Bad connectivity, any other un-handled issues

Wrapping everything under one exception handler only makes sense for that third case (where its an un handled situation that could happen for any  of the **other** number of reasons).

File: src/D2L.Bmx/Aws/AwsClient.cs [hunk 5]
```diff
@@ -13,18 +13,26 @@ internal class AwsClient( IAmazonSecurityTokenService stsClient ) : IAwsClient {
...
+			} );
+
+			return new AwsCredentials(
+				SessionToken: authResp.Credentials.SessionToken,
+				AccessKeyId: authResp.Credentials.AccessKeyId,
+				SecretAccessKey: authResp.Credentials.SecretAccessKey,
+				Expiration: authResp.Credentials.Expiration.ToUniversalTime()
+			);
+		} catch( Exception ex ) {
+			if( ex.Message == "The requested DurationSeconds exceeds the MaxSessionDuration set for this role." ) {
```
Comment (cfbao) [inline src/D2L.Bmx/Aws/AwsClient.cs:32]: this case has a more specific exception type, right?
you can / should use a more specific `catch` clause in such scenarios, e.g.
```csharp
} catch ( AmazonSecurityTokenServiceException ex ) when ( ex.Message == "foo" ) {
	throw new BmxException( "don't do foo", ex );
} catch ( Exception ex ) {
	throw new BmxException( "don't do... idk... this", ex );
}
```

File: src/D2L.Bmx/Aws/AwsClient.cs [hunk 6]
```diff
@@ -13,18 +13,26 @@ internal class AwsClient( IAmazonSecurityTokenService stsClient ) : IAwsClient {
...
+				throw new BmxException( "Duration exceeds the MaxSessionDuration for this role. Lower it in config/parameter", ex );
+			}
+			throw new BmxException( "Error AWS failed to grab credentials", ex );
```
Comment (cfbao) [inline src/D2L.Bmx/Aws/AwsClient.cs:35]: we can rephrase it better? thinking:

"Failed to get credentials from AWS"
or
"Error getting credentials from AWS"

File: src/D2L.Bmx/Okta/OktaApi.cs [hunk 7]
```diff
@@ -49,10 +49,15 @@ internal class OktaApi : IOktaApi {
...
-			new AuthenticateRequest( username, password ),
-			SourceGenerationContext.Default.AuthenticateRequest );
+		HttpResponseMessage resp;
+		try {
+			resp = await _httpClient.PostAsJsonAsync(
+				"authn",
+				new AuthenticateRequest( username, password ),
+				SourceGenerationContext.Default.AuthenticateRequest );
+		} catch( Exception ex ) {
+			throw new BmxException( "Okta authentication request failed. Try checking for a stable connection.", ex );
```
Comment (cfbao) [inline src/D2L.Bmx/Okta/OktaApi.cs:59]: should probably reword this to be broader. bad network connection isn't the only reason this can fail. It can also be Okta's fault (e.g. site down. it happened a few times)

File: src/D2L.Bmx/Okta/OktaApi.cs [hunk 8]
```diff
@@ -73,17 +78,19 @@ internal class OktaApi : IOktaApi {
...
+			var response = await _httpClient.PostAsJsonAsync(
 			$"authn/factors/{factorId}/verify",
 			new IssueMfaChallengeRequest( stateToken ),
 			SourceGenerationContext.Default.IssueMfaChallengeRequest );
 
-		if( !response.IsSuccessStatusCode ) {
-			throw new BmxException( "Error sending MFA challenge" );
+			response.EnsureSuccessStatusCode();
+		} catch( Exception ex ) {
+			throw new BmxException( "Error sending MFA challenge", ex );
```
Comment (cfbao) [inline src/D2L.Bmx/Okta/OktaApi.cs:93]: I know you only kept the original message here, but I'm thinking we can reword this to be more user friendly. e.g.

> Error initiating MFA with Okta

maybe? is this an accurate description?

File: src/D2L.Bmx/Okta/OktaApi.cs [hunk 9]
```diff
@@ -95,41 +102,55 @@ string challengeResponse
...
-			SourceGenerationContext.Default.VerifyMfaChallengeResponseRequest );
-
+		HttpResponseMessage resp;
+		try {
+			resp = await _httpClient.PostAsJsonAsync(
+				$"authn/factors/{factorId}/verify",
+				request,
+				SourceGenerationContext.Default.VerifyMfaChallengeResponseRequest );
+		} catch( Exception ex ) {
+			throw new BmxException( "Error verifying Okta MFA challenge response. Try checking for a stable connection.", ex );
```
Comment (cfbao) [inline src/D2L.Bmx/Okta/OktaApi.cs:112]: similarly here (and below). the message should be reworded to cover broader possible causes

File: src/D2L.Bmx/Okta/OktaApi.cs [hunk 10]
```diff
@@ -95,41 +102,55 @@ string challengeResponse
...
+		}
 		var authnResponse = await JsonSerializer.DeserializeAsync(
```
Comment (cfbao) [inline src/D2L.Bmx/Okta/OktaApi.cs:114]: a try catch around json deserialization would be good too. possible for Okta to return invalid responses.
similarly in a few other places

**Review by cfbao — APPROVED** [review]
Looks good!

(btw I also updated the PR title "Octa" -> "Okta")
